### PR TITLE
storage: skip the lock and rollback record in scan_latest_user_keys

### LINKED ISSUE #15219: scan_latest_user_keys should skip the type of non-user write
## Bug Report

`scan_latest_user_keys` is used to scan and get the latest user keys during the flashback progress, currently, it doesn't consider the key's write record which may cause the following situation:

1. Flashback in phase 1 uses `scan_latest_user_keys` to scan and get the first user key `k1` which has the latest write record with the type `WriteType::Lock` or `WriteType::Rollback`.
2. Flashback prewrite this key.
3. Compaction Filter removes `k1`'s latest record since it's a non-user write.
4. Flashback in phase 2 uses `scan_latest_user_keys` to scan and get the first user key `k2` which doesn't have any lock or commit record with `flashback_commit_ts`.
5. Flashback fails.

`scan_latest_user_keys` should skip the type of non-user write to ensure we can always get the same key in the 2 different phases.

### PR DESCRIPTION
**Pull request by JmPotato**
### What is changed and how it works?

Issue Number: Close #15219.

What's Changed:

```commit-message
Skip the `WriteType:Lock` and `WriteType::Rollback` record in `scan_latest_user_keys`.
```

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
None.
```

### CODE REVIEW

File: src/storage/mvcc/reader/reader.rs [hunk 1]
```diff
@@ -636,6 +639,13 @@ impl<S: EngineSnapshot> MvccReader<S> {
             }
             let commit_ts = key.decode_ts()?;
             let user_key = key.truncate_ts()?;
+            // Skip the key if its latest write type is `WriteType::Lock` or
+            // `WriteType::Rollback`.
+            let write = WriteRef::parse(cursor.value(&mut self.statistics.write))?;
+            if write.write_type != WriteType::Put && write.write_type != WriteType::Delete {
```
[the remark this answers is not in the document]
Comment (JmPotato) [reply src/storage/mvcc/reader/reader.rs]: Fixed. PTAL again.

[the remark this answers is not in the document]
Comment (HuSharp) [reply src/storage/mvcc/reader/reader.rs]: Do you need to update the comment as well?

File: src/storage/mvcc/reader/reader.rs [hunk 2]
```diff
@@ -594,7 +594,10 @@ impl<S: EngineSnapshot> MvccReader<S> {
         Ok((locks, has_remain))
     }
 
-    /// Scan the writes to get all the latest user keys. The return type is:
+    /// Scan the writes to get all the latest user keys. This scan will skip
+    /// `WriteType::Lock` and `WriteType::Rollback`, only return the key that
+    /// has a latest `WriteType::Put` or `WriteType::Delete` record. The return
+    /// type is:
```
Comment (HuSharp) [inline src/storage/mvcc/reader/reader.rs:600]: need to update as well

Comment (JmPotato) [reply src/storage/mvcc/reader/reader.rs:600]: I think this part covers all types already?

Comment (HuSharp) [reply src/storage/mvcc/reader/reader.rs:600]: Well, it makes sense for now :)

**Review by HuSharp — APPROVED** [review]
LGTM!

**Comment by v01dstar** [comment]
After revisiting the code, it seems to me that "latest user key scanning" is just for finding keys that existed before flashback version. `flashback_to_version_write()` is the function that finds the desired versions and rewrite them. So, feels like, in addition to this change, we also need to change this line `let old_write = reader.get_write(&key, flashback_version, None)?;`, instead of getting the `Write` record regardlessly, we probably also have to get either `Write::Delete` or `Write::Put` there.

**Comment by JmPotato** [comment]
> After revisiting the code, it seems to me that "latest user key scanning" is just for finding keys that existed before flashback version. `flashback_to_version_write()` is the function that finds the desired versions and rewrite them. So, feels like, in addition to this change, we also need to change this line `let old_write = reader.get_write(&key, flashback_version, None)?;`, instead of getting the `Write` record regardlessly, we probably also have to get either `Write::Delete` or `Write::Put` there.

`get_write` has already handled this case, it will only return `Write::Delete` and `Write::Put` records. You can check the internal implementation of it.

**Comment by v01dstar** [comment]
> After revisiting the code, it seems to me that "latest user key scanning" is just for finding keys that existed before flashback version. `flashback_to_version_write()` is the function that finds the desired versions and rewrite them. So, feels like, in addition to this change, we also need to change this line `let old_write = reader.get_write(&key, flashback_version, None)?;`, instead of getting the `Write` record regardlessly, we probably also have to get either `Write::Delete` or `Write::Put` there.

This is based on my understanding about "flashback", i.e. it should be like reading with an old timestamp (stale read).

**Review by v01dstar — APPROVED** [review]
LGTM

**Review by Connor1996 — APPROVED** [review]
LGTM

**Pull request merged**